abort nccl communicators before throwing operation timed out

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31128 abort nccl communicators before throwing operation timed out**

When operation times out due to some errors that are not detected by nccl communicators, ncclCommWatchdog can not check this time out error and thus can not abort ncclComms accordingly. So explicitly abort ncclComms here before throwing this timed out exception to users, after this, ncclCommWatchdog can detect nccl communicators are aborted and clean up devNCCLCommMap_ accordingly. if throwing timed out excepiton without aborting nccl communicators here, it was observed that CUDA GPU will have 100% utilization and can not run new events successfully.

Differential Revision: [D18928607](https://our.internmc.facebook.com/intern/diff/D18928607/)

zhaojuanmao committed:
abort nccl communicators before throwing operation timed out

When operation times out due to some errors that are not detected by nccl communicators, ncclCommWatchdog can not check this time out error and thus can not abort ncclComms accordingly. So explicitly abort ncclComms here before throwing this timed out exception to users, after this, ncclCommWatchdog can detect nccl communicators are aborted and clean up devNCCLCommMap_ accordingly. if throwing timed out excepiton without aborting nccl communicators here, it was observed that CUDA GPU will have 100% utilization and can not run new events successfully.

Differential Revision: [D18928607](https://our.internmc.facebook.com/intern/diff/D18928607/)

[ghstack-poisoned]

rohan-varma commented on test/test_c10d.py:
```diff
@@ -3171,9 +3171,12 @@ def _test_nccl_errors_blocking(self, func):
+            try:
                 # Operation would time out in blocking mode.
                 work.wait()
+            except Exception as e:
```
nit: are we not able to do self.assertRaisesRegex and keep the more  specific exception type?

zhaojuanmao replied:
I'm intentionally try exception here to run a cuda operation after catching the exception, so that we can reproduce the nccl stuck issue (without fixing in this diff)

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -201,6 +201,18 @@ void ProcessGroupNCCL::WorkNCCL::synchronize() {
+        // communicators are aborted and clean up devNCCLCommMap_ accordingly.
+        // if throwing timed out excepiton without aborting nccl communicators
+        // here, it was observed that CUDA GPU will have 100% utilization and
+        // can not run new events successfully.
```
Did we verify that the watchdog detects an issue with the communicator and successfully removes the communicator? I think we should add a test in `ProcessGroupNCCLErrorsTest` to ensure that the communicators are cleared from the cache.

zhaojuanmao replied:
yes, local prints showed watch dog received error and clean it up accordingly. OK, will think about how to verify communicators are cleared

zhaojuanmao replied:
added a test for timedout error in ProcessGroupNCCLErrorsTest, it validated that the communicators are cleared from the cache as well

pritamdamania87 commented on test/test_c10d.py:
```diff
@@ -3171,9 +3171,12 @@ def _test_nccl_errors_blocking(self, func):
                 work.wait()
+            except Exception as e:
+                self.assertEqual("Operation timed out!", str(e))
+                a = torch.rand(10).cuda(self.rank)
```
What happens if we try another allreduce operation here?

zhaojuanmao replied:
I think it will time out unless we want to init process group again with rank 1 and both of rank 0 and rank 1 ran all reduce again

zhaojuanmao replied:
let me try that

zhaojuanmao replied:
I tried a few things, run another allreduce will just stuck, I think the two processes are not in the clean sync state to recover in our unit tests. Did not successfully to make these two processes in a clean sync state to rerun all reduce so far (PET is doing that)

isunjin commented on test/test_c10d.py:
```diff
@@ -3171,9 +3171,12 @@ def _test_nccl_errors_blocking(self, func):
                 work.wait()
+            except Exception as e:
+                self.assertEqual("Operation timed out!", str(e))
+                a = torch.rand(10).cuda(self.rank)
```
what is the purpose for this? are you verifying GPU is not stuck at 100% and memory copy is still work. Could you add comments for this?

zhaojuanmao replied:
yes, that  is the main purpose, will add comment

zhaojuanmao committed:
Update on "abort nccl communicators before throwing operation timed out"


When operation times out due to some errors that are not detected by nccl communicators, ncclCommWatchdog can not check this time out error and thus can not abort ncclComms accordingly. So explicitly abort ncclComms here before throwing this timed out exception to users, after this, ncclCommWatchdog can detect nccl communicators are aborted and clean up devNCCLCommMap_ accordingly. if throwing timed out excepiton without aborting nccl communicators here, it was observed that CUDA GPU will have 100% utilization and can not run new events successfully.

Differential Revision: [D18928607](https://our.internmc.facebook.com/intern/diff/D18928607/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 2832e87:
* **2/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2832e87dad3327ec62935e79ce86b54db2119cc4).

### 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3913700) pytorch_linux_xenial_py3_clang5_asan_test (1/2)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=72722887))
```
Dec 12 01:54:56   File "test_nn.py", line 56, in <module>
Dec 12 01:54:56     import hypothesis_utils as hu
Dec 12 01:54:56   File "/var/lib/jenkins/workspace/test/hypothesis_utils.py", line 9, in <module>
Dec 12 01:54:56     from hypothesis.searchstrategy import SearchStrategy
Dec 12 01:54:56 ModuleNotFoundError: No module named 'hypothesis.searchstrategy'
Dec 12 01:54:56 Traceback (most recent call last):
Dec 12 01:54:56   File "test/run_test.py", line 456, in <module>
Dec 12 01:54:56     main()
Dec 12 01:54:56   File "test/run_test.py", line 449, in main
Dec 12 01:54:56     raise RuntimeError(message)
Dec 12 01:54:56 RuntimeError: test_nn failed!
Dec 12 01:54:56 =================== sccache compilation log ===================
Dec 12 01:54:56 + cleanup
Dec 12 01:54:56 + retcode=1
Dec 12 01:54:56 + set +x
Dec 12 01:54:56 =========== If your build fails, please take a look at the log above for possible reasons ===========
Dec 12 01:54:56 Compile requests                  7
Dec 12 01:54:56 Compile requests executed         6
Dec 12 01:54:56 Cache hits                        0
Dec 12 01:54:56 Cache misses                      6
Dec 12 01:54:56 Cache timeouts                    0
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3913692) pytorch_macos_10_13_py3_test (2/2)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=72723598))
```
Dec 11 17:55:39   File "test_fake_quant.py", line 7, in <module>
Dec 11 17:55:39     import hypothesis_utils as hu
Dec 11 17:55:39   File "/Users/distiller/project/test/hypothesis_utils.py", line 9, in <module>
Dec 11 17:55:39     from hypothesis.searchstrategy import SearchStrategy
Dec 11 17:55:39 ModuleNotFoundError: No module named 'hypothesis.searchstrategy'
Dec 11 17:55:39 Traceback (most recent call last):
Dec 11 17:55:39   File "test/run_test.py", line 456, in <module>
Dec 11 17:55:39     main()
Dec 11 17:55:39   File "test/run_test.py", line 449, in main
Dec 11 17:55:39     raise RuntimeError(message)
Dec 11 17:55:39 RuntimeError: test_fake_quant failed!
Dec 11 17:55:39 + cleanup
Dec 11 17:55:39 + retcode=1
Dec 11 17:55:39 + set +x
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

zhaojuanmao committed:
Update on "abort nccl communicators before throwing operation timed out"


When operation times out due to some errors that are not detected by nccl communicators, ncclCommWatchdog can not check this time out error and thus can not abort ncclComms accordingly. So explicitly abort ncclComms here before throwing this timed out exception to users, after this, ncclCommWatchdog can detect nccl communicators are aborted and clean up devNCCLCommMap_ accordingly. if throwing timed out excepiton without aborting nccl communicators here, it was observed that CUDA GPU will have 100% utilization and can not run new events successfully.

Differential Revision: [D18928607](https://our.internmc.facebook.com/intern/diff/D18928607/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "abort nccl communicators before throwing operation timed out"


When operation times out due to some errors that are not detected by nccl communicators, ncclCommWatchdog can not check this time out error and thus can not abort ncclComms accordingly. So explicitly abort ncclComms here before throwing this timed out exception to users, after this, ncclCommWatchdog can detect nccl communicators are aborted and clean up devNCCLCommMap_ accordingly. if throwing timed out excepiton without aborting nccl communicators here, it was observed that CUDA GPU will have 100% utilization and can not run new events successfully.

Differential Revision: [D18928607](https://our.internmc.facebook.com/intern/diff/D18928607/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "abort nccl communicators before throwing operation timed out"


When operation times out due to some errors that are not detected by nccl communicators, ncclCommWatchdog can not check this time out error and thus can not abort ncclComms accordingly. So explicitly abort ncclComms here before throwing this timed out exception to users, after this, ncclCommWatchdog can detect nccl communicators are aborted and clean up devNCCLCommMap_ accordingly. if throwing timed out excepiton without aborting nccl communicators here, it was observed that CUDA GPU will have 100% utilization and can not run new events successfully.

Differential Revision: [D18928607](https://our.internmc.facebook.com/intern/diff/D18928607/)

[ghstack-poisoned]

pritamdamania87 reviewed (approved):
Looks good to me! Thanks for fixing this quickly!

zhaojuanmao commented:
@pytorchbot please retest

zhaojuanmao commented:
@pytorchbot retest please

zhaojuanmao commented:
@pytorchbot rebase

zhaojuanmao committed:
Update on "abort nccl communicators before throwing operation timed out"


When operation times out due to some errors that are not detected by nccl communicators, ncclCommWatchdog can not check this time out error and thus can not abort ncclComms accordingly. So explicitly abort ncclComms here before throwing this timed out exception to users, after this, ncclCommWatchdog can detect nccl communicators are aborted and clean up devNCCLCommMap_ accordingly. if throwing timed out excepiton without aborting nccl communicators here, it was observed that CUDA GPU will have 100% utilization and can not run new events successfully.

Differential Revision: [D18928607](https://our.internmc.facebook.com/intern/diff/D18928607/)

[ghstack-poisoned]

zhaojuanmao commented:
failures are not related, landing soon
